server/selection.go

```diff
@@ -190,8 +195,11 @@ func (s *MinLSSelector) selectUnknownSession() *BroadcastSession {
 	// The greater the stake weight of a session, the more likely that it will be selected because subtracting its stake weight from r
 	// will result in a value <= 0
 	for i, sess := range s.unknownSessions {
+		addr := ethcommon.BytesToAddress(sess.OrchestratorInfo.TicketParams.Recipient)
 		// If we could not fetch the stake weight for addrs[i] then its stake weight defaults to 0
-		r -= stakes[addrs[i]]
+		// The stake weight for a session is the stake weight of the address divided by the # of sessions the
+		// address is associated with
+		r -= (stakes[addr] / int64(addrCount[addr]))
```

Comment:
Am I correct in assuming that rounding might only a problem here for tiny amounts of stake, which isn't typically a problem in practice? (So it's safe to ignore the small bias from rounding down.)

Reply:
If  `stakes[addr] < addrCount[addr]` we end  up with a stake weight of 0 , I think it's valid to think about how we should handle that. Should we penalise the node operator for running too many nodes with a very low stake which would not warrant running many nodes because the chances of being selected  are slim or should we default to `stakes[addr]` when it is  smaller than the `addrCount`?

Reply:
I updated the stake weight division in 1502b62.

For each session associated with an address, the session is assigned the remaining stake weight for the address (`stakes[addr]`) divided by the current # of sessions for the address (`addrCount[addr]`). After assigning the stake weight of a session, the session's stake weight is subtracted by the remaining stake weight for the address and the current # of sessions for the address is decremented. The result is that the last session for an address during a `selectUnknownSession()` call will be assigned the remaining stake weight of the address such that the entire stake weight of the address is used.

In the case where `stakes[addr] < addrCount[addr]` at the start of selection, each session would be assigned a stake weight of 0 except for one session which will be assigned a stake weight of `stakes[addr]`. This means that the maximum # of nodes that can be eligible for selection for an operator is `stakes[addr]` (the working set is capped at 8 right now so you can divide stake amongst at most 8 sessions right now). That seems like an ok restriction right now.

Reply:
Nice approach @yondonfu .

If I'm understanding the overall flow correctly, one consequence might be that the stake is weighted towards the last node of `addr` in the `unknownSessions` list. Is the last node also the one that took the longest to respond to `OrchestratorInfo` request? Which may mean the node is under load, far away, network congested, etc.

Not sure how much of an effect this will actually have in practice though; I imagine it's only under specific cases (minuscule amounts of stake + issues with the last O) where we might see issues as the B selects the first node or so from a refreshed list.

Reply:
Yeah the last node of `addr` in `unknownSessions` would either be the node that took the longest to respond during discovery or a node that was suspended, but included in `unknownSessions` because there were insufficient non-suspended nodes that responded during discovery.

In practice, I think this will only matter in the case that you mention with small amounts of stake + issues with the last node. In the interest of avoiding bias though, I pushed up 4a1cc12 which randomly assigns the "leftover" stake for `addr` to one of its sessions.

While working on that commit, I also discovered a bug that could cause zero stake weight sessions to be selected before non-zero stake weight selections. Fixed in 4a1cc12